// nativescript-statusbar/angular/statusbar-directives.js
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var element_registry_1 = require("nativescript-angular/element-registry");
var __1 = require("./../");
var StatusBarComponent = /** @class */ (function () {
    function StatusBarComponent() {
    }
    StatusBarComponent = __decorate([
        core_1.Component({
            selector: "StatusBar",
            template: ""
        })
    ], StatusBarComponent);
    return StatusBarComponent;
}());
exports.StatusBarComponent = StatusBarComponent;
exports.STATUSBAR_DIRECTIVES = [StatusBarComponent];
if (!global.isListViewRegistered) {
    element_registry_1.registerElement('StatusBar', function () { return __1.StatusBar; });
    global.isListViewRegistered = true;
}
var NativeScriptStatusBarModule = /** @class */ (function () {
    function NativeScriptStatusBarModule() {
    }
    NativeScriptStatusBarModule = __decorate([
        core_1.NgModule({
            declarations: [exports.STATUSBAR_DIRECTIVES],
            exports: [exports.STATUSBAR_DIRECTIVES],
            schemas: [
                core_1.NO_ERRORS_SCHEMA
            ]
        })
    ], NativeScriptStatusBarModule);
    return NativeScriptStatusBarModule;
}());
exports.NativeScriptStatusBarModule = NativeScriptStatusBarModule;
