// nativescript/lib/commands/update-platform.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class UpdatePlatformCommand {
    constructor($errors, $options, $platformEnvironmentRequirements, $platformCommandHelper, $platformValidationService, $projectData) {
        this.$errors = $errors;
        this.$options = $options;
        this.$platformEnvironmentRequirements = $platformEnvironmentRequirements;
        this.$platformCommandHelper = $platformCommandHelper;
        this.$platformValidationService = $platformValidationService;
        this.$projectData = $projectData;
        this.allowedParameters = [];
        this.$projectData.initializeProjectData();
    }
    execute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.$platformCommandHelper.updatePlatforms(args, this.$projectData);
        });
    }
    canExecute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!args || args.length === 0) {
                this.$errors.failWithHelp("No platform specified. Please specify platforms to update.");
            }
            _.each(args, arg => {
                const platform = arg.split("@")[0];
                this.$platformValidationService.validatePlatform(platform, this.$projectData);
            });
            for (const arg of args) {
                const [platform, versionToBeInstalled] = arg.split("@");
                const checkEnvironmentRequirementsInput = { platform, options: this.$options };
                if (versionToBeInstalled) {
                    checkEnvironmentRequirementsInput.projectDir = this.$projectData.projectDir;
                    checkEnvironmentRequirementsInput.runtimeVersion = versionToBeInstalled;
                }
                yield this.$platformEnvironmentRequirements.checkEnvironmentRequirements(checkEnvironmentRequirementsInput);
            }
            return true;
        });
    }
}
exports.UpdatePlatformCommand = UpdatePlatformCommand;
$injector.registerCommand("platform|update", UpdatePlatformCommand);
