// nativescript/lib/common/commands/device/list-files.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class ListFilesCommand {
    constructor($devicesService, $stringParameter, $options, $projectData, $errors) {
        this.$devicesService = $devicesService;
        this.$stringParameter = $stringParameter;
        this.$options = $options;
        this.$projectData = $projectData;
        this.$errors = $errors;
        this.allowedParameters = [this.$stringParameter, this.$stringParameter];
    }
    execute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.$devicesService.initialize({ deviceId: this.$options.device, skipInferPlatform: true });
            const pathToList = args[0];
            let appIdentifier = args[1];
            if (!appIdentifier) {
                try {
                    this.$projectData.initializeProjectData();
                }
                catch (err) {
                }
                if (!this.$projectData.projectIdentifiers) {
                    this.$errors.fail("Please enter application identifier or execute this command in project.");
                }
            }
            const action = (device) => __awaiter(this, void 0, void 0, function* () {
                appIdentifier = appIdentifier || this.$projectData.projectIdentifiers[device.deviceInfo.platform.toLowerCase()];
                yield device.fileSystem.listFiles(pathToList, appIdentifier);
            });
            yield this.$devicesService.execute(action);
        });
    }
}
exports.ListFilesCommand = ListFilesCommand;
$injector.registerCommand(["device|list-files", "devices|list-files"], ListFilesCommand);
