// nativescript/lib/services/analytics/analytics-broker.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../common/decorators");
class AnalyticsBroker {
    constructor($analyticsSettingsService, $injector, analyticsLoggingService) {
        this.$analyticsSettingsService = $analyticsSettingsService;
        this.$injector = $injector;
        this.analyticsLoggingService = analyticsLoggingService;
    }
    getGoogleAnalyticsProvider() {
        return __awaiter(this, void 0, void 0, function* () {
            const clientId = yield this.$analyticsSettingsService.getClientId();
            return this.$injector.resolve("googleAnalyticsProvider", { clientId, analyticsLoggingService: this.analyticsLoggingService });
        });
    }
    sendDataForTracking(trackInfo) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const googleProvider = yield this.getGoogleAnalyticsProvider();
                yield googleProvider.trackHit(trackInfo);
            }
            catch (err) {
                this.analyticsLoggingService.logData({ message: `AnalyticsBroker unable to execute action in sendDataForTracking: ${err}`, type: "Error" });
            }
        });
    }
}
__decorate([
    decorators_1.cache()
], AnalyticsBroker.prototype, "getGoogleAnalyticsProvider", null);
exports.AnalyticsBroker = AnalyticsBroker;
