// nativescript/lib/services/livesync/playground/preview-sdk-service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const nativescript_preview_sdk_1 = require("nativescript-preview-sdk");
const pako = require("pako");
class PreviewSdkService {
    constructor($config, $httpClient, $logger, $previewDevicesService, $previewAppLogProvider, $previewSchemaService) {
        this.$config = $config;
        this.$httpClient = $httpClient;
        this.$logger = $logger;
        this.$previewDevicesService = $previewDevicesService;
        this.$previewAppLogProvider = $previewAppLogProvider;
        this.$previewSchemaService = $previewSchemaService;
        this.messagingService = null;
        this.instanceId = null;
    }
    getQrCodeUrl(options) {
        const { projectDir, useHotModuleReload } = options;
        const schema = this.$previewSchemaService.getSchemaData(projectDir);
        const hmrValue = useHotModuleReload ? "1" : "0";
        const result = `${schema.name}://boot?instanceId=${this.instanceId}&pKey=${schema.publishKey}&sKey=${schema.subscribeKey}&template=play-ng&hmr=${hmrValue}`;
        return result;
    }
    initialize(projectDir, getInitialFiles) {
        return __awaiter(this, void 0, void 0, function* () {
            const initConfig = this.getInitConfig(projectDir, getInitialFiles);
            this.messagingService = new nativescript_preview_sdk_1.MessagingService();
            this.instanceId = yield this.messagingService.initialize(initConfig);
        });
    }
    applyChanges(filesPayload) {
        return new Promise((resolve, reject) => {
            this.messagingService.applyChanges(this.instanceId, filesPayload, err => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
    stop() {
        this.messagingService.stop();
    }
    getInitConfig(projectDir, getInitialFiles) {
        const schema = this.$previewSchemaService.getSchemaData(projectDir);
        return {
            pubnubPublishKey: schema.publishKey,
            pubnubSubscribeKey: schema.subscribeKey,
            msvKey: schema.msvKey,
            msvEnv: this.$config.PREVIEW_APP_ENVIRONMENT,
            callbacks: this.getCallbacks(),
            getInitialFiles,
            previewAppStoreId: schema.previewAppStoreId,
            previewAppGooglePlayId: schema.previewAppId
        };
    }
    getCallbacks() {
        return {
            onLogSdkMessage: (log) => {
                this.$logger.trace("Received onLogSdkMessage message: ", log);
            },
            onLogMessage: (log, deviceName, deviceId) => {
                this.$previewAppLogProvider.logData(log, deviceName, deviceId);
            },
            onRestartMessage: () => {
                this.$logger.trace("Received onRestartMessage event.");
            },
            onUncaughtErrorMessage: () => {
                this.$logger.warn("The Preview app has terminated unexpectedly. Please run it again to get a detailed crash report.");
            },
            onConnectedDevicesChange: (connectedDevices) => ({}),
            onDeviceConnectedMessage: (deviceConnectedMessage) => ({}),
            onDeviceConnected: (device) => ({}),
            onDevicesPresence: (devices) => this.$previewDevicesService.updateConnectedDevices(devices),
            onSendingChange: (sending) => ({}),
            onBiggerFilesUpload: (filesContent, callback) => __awaiter(this, void 0, void 0, function* () {
                const gzippedContent = Buffer.from(pako.gzip(filesContent));
                const byteLength = filesContent.length;
                if (byteLength > PreviewSdkService.MAX_FILES_UPLOAD_BYTE_LENGTH) {
                    this.$logger.warn("The files to upload exceed the maximum allowed size of 15MB. Your app might not work as expected.");
                }
                const playgroundUploadResponse = yield this.$httpClient.httpRequest({
                    url: this.$config.UPLOAD_PLAYGROUND_FILES_ENDPOINT,
                    method: "POST",
                    body: gzippedContent,
                    headers: {
                        "Content-Encoding": "gzip",
                        "Content-Type": "text/plain"
                    }
                });
                const responseBody = JSON.parse(playgroundUploadResponse.body);
                const location = responseBody && responseBody.location;
                callback(location, playgroundUploadResponse.error);
            })
        };
    }
}
PreviewSdkService.MAX_FILES_UPLOAD_BYTE_LENGTH = 15 * 1024 * 1024;
exports.PreviewSdkService = PreviewSdkService;
$injector.register("previewSdkService", PreviewSdkService);
